herd/AArch64Sem.ml

```diff
@@ -1055,30 +1028,100 @@ module Make
    + ma abstracted for all variants
  *)
 
-      let to_perms str sz = str ^ if sz = MachSize.S128 then "_c" else ""
-
-      let apply_mv mop mv = fun ac ma -> mop ac ma mv
-
       let is_this_reg rA e =
         match E.location_reg_of e with
         | None -> false
         | Some rB -> AArch64.reg_compare rA rB=0
 
+      (*  memtag faults *)
+
+      let lift_fault_memtag mfault mm dir ii =
+        if has_handler ii then
+          fun ma -> M.bind_ctrldata ma (fun _ -> mfault >>! B.Fault dir)
+        else
+          let open Precision in
+          match C.precision,dir with
+          | (Fatal,_)|(LoadsFatal,(Dir.R)) ->
+             fun ma ->  ma >>*= (fun _ -> mfault >>! B.Exit)
+          | (Handled,_)|(LoadsFatal,Dir.W) ->
+             fun ma ->
+             let (>>) = M.bind_ctrl_first_outputs in
+             let ma = ma >> (fun a -> mfault >>! a) in
+             mm ma >>! B.Next []
+          | Skip,_ ->
+             fun _ ->
+               Warn.fatal "Memtag extension has no 'Skip' fault handling mode"
+
+      let lift_memtag_phy rA dir updatedb mop ma an ii =
+        let lift_tag =
+          let mphy = fun ma a_virt ->
+            M.delay_kont "4" ma
+              (fun a_phy ma ->
+                 delayed_check_tags a_virt (Some a_phy) ma ii
+                   (fun ma -> ma >>= M.ignore >>= B.next1T)
+                   (fun ma -> ma >>! B.Fault dir)) in
+          lift_kvm Dir.R false true mop ma an ii mphy in
+        let phy_addr_reg = AArch64.Internal 0 in
+        let lift_data =
+          let mphy = fun ma _ -> ma >>= fun a -> M.unitT (B.Next [(phy_addr_reg, a)]) in
+          lift_kvm dir updatedb false mop ma an ii mphy in
+        let mphy mtag mdata =
+          let get_phy_addr a_virt = function
+            | B.Next bds -> begin
+                try M.unitT (List.assoc phy_addr_reg bds)
+                with Not_found -> M.unitT a_virt
+              end
+            | _ -> Warn.fatal "Cannot handle argument to get_phy_addr" in
+          let ma = ma >>= fun a_virt -> M.aslseq mtag (fun _ -> mdata) >>=
```

Comment:
I agree with you, I was very close to reimplementing `aslseq` because I thought "aslseq is for ASL stuff". It was only when I started implementing `para_output_right` that I found that it was already implemented. All that to say, I think `aslseq` is more general purpose than its name :)